original/man/man3/bnlearn.3
.TH "bnlearn" 3 "28 Jul 2003" "VFML" \" -*- nroff -*-
.ad l
.nh
.SH NAME
bnlearn \- 
.SH "Detailed Description"
.PP 
Learns the structure of a BeliefNet from a data set. Designed to be easily modified. 

Learns the structure and parameters of a Bayesian network using the standard method. The structural prior is set by using P(S) = kappa ^ (symetric difference between the net and the prior net). The parameter prior is K2. This is basically a wrapper around the code in bnlearn-engine.h. This program is a stripped down version of the vfbn learner, done so that the code would be clearer and easier to modify.
.PP
bnlearn can load training data into RAM, if space is available, or it can scan data repeatedly from disk.
.PP
See the documentation for the bnlearn-engine for even more information about the parameters and their meanings.
.PP
bnlearn takes input and does output in \fCc4.5 format\fP. It expects to find the files \fC<stem>.names\fP and \fC<stem>.data\fP.
.PP
\fBThanks\fP
.RS 4
to Matthew Richardson for making substantial optimizations to the bnlearn program. 
.RE
.PP
.PP
\fBWish List\fP
.RS 4
A version of this program that is intelligent about dealing with unobserved (or partially observed) variables. 
.RE
.PP
.SS "Arguments"
.PP
.IP "\(bu" 2
-f 'filestem'
.IP "  \(bu" 4
Set the name of the dataset (default DF)
.PP

.IP "\(bu" 2
-source 'dir'
.IP "  \(bu" 4
Set the source data directory (default '.')
.PP

.IP "\(bu" 2
-startFrom 'filename'
.IP "  \(bu" 4
Use net in filename as starting point, must be BIF file (default start from empty net)
.PP

.IP "\(bu" 2
-outputTo 'filename'
.IP "  \(bu" 4
Output the learned net to filename
.PP

.IP "\(bu" 2
-limitMegs 'count'
.IP "  \(bu" 4
Limit dynamic memory allocation to 'count' megabytes, don't consider networks that take too much space (default no limit)
.PP

.IP "\(bu" 2
-limitMinutes 'count'
.IP "  \(bu" 4
Limit the run to 'count' minutes then return current model (default no limit)
.PP

.IP "\(bu" 2
-noCacheTrainSet
.IP "  \(bu" 4
Repeatedly read training data from disk and do not cache it in RAM (default read data into RAM at beginning of run)
.PP

.IP "\(bu" 2
-stdin
.IP "  \(bu" 4
Reads training examples from stdin instead of from stem.data causes a 2 second delay to help give input time to setup (default off, not compatable with -noCacheTrainSet)
.PP

.IP "\(bu" 2
-noReverse
.IP "  \(bu" 4
Doesn't reverse links to make nets for next search step (default reverse links)
.PP

.IP "\(bu" 2
-parametersOnly
.IP "  \(bu" 4
Only estimate parameters for current structure, no other learning
.PP

.IP "\(bu" 2
-seed 's'
.IP "  \(bu" 4
Seed for random numbers (default random)
.PP

.IP "\(bu" 2
-maxSearchSteps 'num'
.IP "  \(bu" 4
Limit to 'num' search steps (default no max).
.PP

.IP "\(bu" 2
-maxParentsPerNode 'num'
.IP "  \(bu" 4
Limit each node to 'num' parents (default no max).
.PP

.IP "\(bu" 2
-maxParameterGrowthMult 'mult'
.IP "  \(bu" 4
Limit net to 'mult' times starting # of parameters (default no max).
.PP

.IP "\(bu" 2
-maxParameterCount 'count'
.IP "  \(bu" 4
Limit net to 'count' parameters (default no max).
.PP

.IP "\(bu" 2
-kappa 'kappa'
.IP "  \(bu" 4
The structure prior penalty for batch (0 - 1), 1 is no penalty (default 0.5)
.PP

.IP "\(bu" 2
-v
.IP "  \(bu" 4
Can be used multiple times to increase the debugging output
.PP

.IP "\(bu" 2
-h
.IP "  \(bu" 4
Run bnlearn -h for a list of the arguments and their meanings.
.PP

.PP

.PP
.SH SYNOPSIS
.br
.PP
.SH "Author"
.PP 
Generated automatically by Doxygen for VFML from the source code.


src/learners/bnlearn/bnlearn.c
#include "vfml.h"
#include <stdio.h>
#include <string.h>

#include <assert.h>

#include <sys/times.h>
#include <unistd.h>
#include <time.h>
#include <math.h>

static void _printUsage(char  *name) {
   printf("%s : Learn the structure (and parameters) of a Belief Net\n", name);

   printf("-f <filestem>\t Set the name of the dataset (default DF)\n");
   printf("-source <dir>\t Set the source data directory (default '.')\n");
   printf("-startFrom <filename>\t use net in <filename> as starting point,\n\t\t must be BIF file (default start from empty net)\n");
   printf("-outputTo <filename>\t output the learned net to <filename>\n");
   printf("-limitMegs <count>\t Limit dynamic memory allocation to 'count'\n\t\t megabytes, don't consider networks that take too much\n\t\t space (default no limit)\n");
   printf("-limitMinutes <count>\t Limit the run to <count> minutes then\n\t\t return current model (default no limit)\n");
   printf("-noCacheTrainSet \t Repeatedly read training data from disk and do not\n\t\t cache it in RAM (default read data into RAM\n\t\t at beginning of run)\n");
   printf("-stdin \t\t Reads training examples from stdin instead of from\n\t\t <stem>.data causes a 2 second delay to help give\n\t\t input time to setup (default off, not compatable\n\t\t with -noCacheTrainSet)\n");
   printf("-noReverse \t Doesn't reverse links to make nets for next search\n\t\t step (default reverse links)\n");
   printf("-parametersOnly\t Only estimate parameters for current\n\t\t structure, no other learning\n");
   printf("-seed <s>\t Seed for random numbers (default random)\n");
   printf("-maxSearchSteps <num>\tLimit to <num> search steps (default no max).\n");
   printf("-maxParentsPerNode <num>\tLimit each node to <num> parents\n\t\t (default no max).\n");
   printf("-maxParameterGrowthMult <mult>\tLimit net to <mult> times starting\n\t\t # of parameters (default no max).\n");
   printf("-maxParameterCount <count>\tLimit net to <count> parameters\n\t\t (default no max).\n");
   printf("-kappa <kappa>  the structure prior penalty for batch (0 - 1), 1 is\n\t\t no penalty (default 0.5)\n");
   printf("-v\t\t Can be used multiple times to increase the debugging output\n");
}


static void _processArgs(int argc, char *argv[], BNLearnParams *params) {
   int i;
   int cacheTrainSet=1;
   char filename[2000];
   ExampleSpecPtr es;
   char *sourceDirectory=".", *fileStem="DF";
   char *inputNetFilename=NULL;
   int useStdin=0;
   FILE *in;

   /* HERE on the ones that use the next arg make sure it is there */
   for(i = 1 ; i < argc ; i++) {
      if(!strcmp(argv[i], "-f")) {
         fileStem = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-startFrom")) {
         inputNetFilename = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-outputTo")) {
         params->gOutputNetFilename = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-source")) {
         sourceDirectory = argv[i+1];
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-v")) {
         DebugSetMessageLevel(DebugGetMessageLevel() + 1);
      } else if(!strcmp(argv[i], "-h")) {
         _printUsage(argv[0]);
         exit(0);
      } else if(!strcmp(argv[i], "-kappa")) {
         sscanf(argv[i+1], "%lf", &params->gKappa);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-limitMegs")) {
         sscanf(argv[i+1], "%ld", &params->gLimitBytes);
         params->gLimitBytes *= 1024 * 1024;
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-limitMinutes")) {
         sscanf(argv[i+1], "%lf", &params->gLimitSeconds);
         params->gLimitSeconds *= 60;
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-maxSearchSteps")) {
         sscanf(argv[i+1], "%d", &params->gMaxSearchSteps);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-maxParentsPerNode")) {
         sscanf(argv[i+1], "%d", &params->gMaxParentsPerNode);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-maxParameterGrowthMult")) {
         sscanf(argv[i+1], "%d", &params->gMaxParameterGrowthMult);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-maxParameterCount")) {
         sscanf(argv[i+1], "%ld", &params->gMaxParameterCount);
         /* ignore the next argument */
         i++;
      } else if(!strcmp(argv[i], "-noCacheTrainSet")) {
         cacheTrainSet=0;
      } else if(!strcmp(argv[i], "-stdin")) {
         sleep(2);
         useStdin=1;
      } else if(!strcmp(argv[i], "-parametersOnly")) {
         params->gOnlyEstimateParameters = 1;
      } else if(!strcmp(argv[i], "-noReverse")) {
         params->gNoReverse = 1;
      } else if(!strcmp(argv[i], "-seed")) {
         sscanf(argv[i+1], "%d", &params->gSeed);
         /* ignore the next argument */
         i++;
      } else {
         printf("Unknown argument: %s.  use -h for help\n", argv[i]);
         exit(0);
      }
   }

   // Verify params
   if(useStdin && (cacheTrainSet == 0)) {
      DebugError(1, "Can not use -stdin with -noCacheTrainSet\n");
   }
   // Set up the input data
   sprintf(filename, "%s/%s.names", sourceDirectory, fileStem);
   es = ExampleSpecRead(filename);
   DebugError(es == 0, "Unable to open the .names file");

   sprintf(filename, "%s/%s.data", sourceDirectory, fileStem);
   in = fopen(filename, "r");
   DebugError(in == 0, "Unable to open the .data file");

   if (cacheTrainSet) {
      params->gDataMemory = ExamplesRead(in,es);
      DebugError(!params->gDataMemory, "Unable to read the .data file");
   }
   else {
      params->gDataFile = in;
   }
   DebugMessage(1, 1, "Stem: %s\n", fileStem);
   DebugMessage(1, 1, "Source: %s\n", sourceDirectory);
   DebugMessage(useStdin, 1, "Reading examples from standard in.\n");

   // Set up the input network
   if (inputNetFilename) {
      params->gInputNetFile = fopen(inputNetFilename, "r");
      DebugError(!params->gInputNetFile, "Unable to reading input network");
   }
   else {
      params->gInputNetEmptySpec = es;
   }

}


int main(int argc, char *argv[]) {
   BNLearnParams *params = BNLearn_NewParams();
   params->gOutputNetToMemory = 1;
   _processArgs(argc, argv, params);   

   BNLearn(params);

   printf("Final net:\n");
   BNWriteBIF(params->gOutputNetMemory, stdout);
   BNPrintStats(params->gOutputNetMemory);

   BNLearn_FreeParams(params);
   return 0;
}
